Remove deactivated users from anonymous groups.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to remove deactivated users from anonymous groups.
- Second commit is a refactor which Shubham suggested for the first commit. I am not sure whether that will help in readability as we need to pass many variables to the extracted function, so have kept it in a separate commit for now.
- Third commit adds the migration.

Fixes: <!-- Issue link, or clear description.--> https://chat.zulip.org/#narrow/channel/3-backend/topic/updating.20group.20permission.20settings

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 committed:
users: Remove users from anonymous groups on deactivation.

This commit adds code to remove deactivated users from
anonymous groups used for different permission settings.
We need to remove deactivated users from anonymous groups
so that we can easily handle updating them where we
check old values as well to avoid race conditions and
we are fine with those users not having the permission
after we reactivate them.

If the anonymous group for that setting contains only the
deactivated user and no other members or subgroups, then
the setting is set to "role:nobody" group. If that
setting cannot be set to "role:nobody" group, then we
raise an error.

As a result of this change, we now send setting update
events as well to all the users for all the settings that
are changed due to the user being removed from anonymous
group. Previously these events were only sent to users who
would not receive the user deactivation event.

We also store details about the permissions being removed
for the deactivated user in the USER_DEACTIVATED type
audit log entry "removed_permissions" field in extra_data
object, so that it is easy to see the permissions that the
user had in case we want to re-assign those permissions to
the user when reactivating them.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -484,15 +484,12 @@ def do_deactivate_user(
+        if _cascade:
+            bot_profiles = get_active_bots_owned_by_user(user_profile)
+            for profile in bot_profiles:
+                do_deactivate_user(profile, _cascade=False, acting_user=acting_user)
```
Not sure why this was not inside a transaction before, but should be fine inlcuding this inside the transaction.

sahil839 replied:
It is not important. I thought would be better to avoid deactivating just bot and not users in case of any failure. Will remove this commit once we decide how to move forward.

sahil839 commented on zerver/management/commands/deactivate_user.py:
```diff
@@ -43,5 +47,14 @@ def handle(self, *args: Any, **options: Any) -> None:
+        except JsonableError as e:
+            raise CommandError(e.msg)
+
+        do_deactivate_user(
```
Need to do this for other management commands as well like deleting user.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -516,6 +745,12 @@ def do_deactivate_user(
             },
         )
+
+        if group_setting_updates is not None:
```
We would also need to remove deactivated bots here from the anonymous groups. So, would need to modify the `check_group_permission_updates_for_deactivating_user` function to handle multiple user_profile.

sahil839 commented on zerver/views/users.py:
```diff
@@ -143,7 +144,11 @@ def deactivate_user_own_backend(request: HttpRequest, user_profile: UserProfile)
-    do_deactivate_user(user_profile, acting_user=user_profile)
+    group_setting_updates = check_group_permission_updates_for_deactivating_user(user_profile)
+
+    do_deactivate_user(
```
Users are also deactivated from LDAP and SCIM code, so need to do that also, but would be good to get an initial review on the approach once.

sahil839 commented:
This is not completed yet and we might be able to do some code deduplication as well, but would be good to get an initial review on approach.

timabbott commented:
I'm not 100% confident we have the right product direction for this change. Let me think about it with @alya and get back to you.

timabbott commented:
I'll try to post something longer in #design, but I think the key thing we want here is an undo mechanism, at the data layer. Concretely what that means is I think we should put something in `RealmAuditLog` for the user deactivation that records in some nicely structured format what permissions were lost with the deactivation. We can't reference anonymous user groups in doing that, since those get deleted, so I think it'd be a set of objects with (say) keys `type: "group` `group_id`, `setting_name` if it's setting on a NamedUserGroup object.

That way, if there's a bug with say LDAP synchronization that deactivates everyone, there's a straightforward way to restore the permissions. (And we could later have the reactivation UI present the opportunity to do it).

I guess it won't work forever, in that we make new settings and rename them, but that can probably be managed. And I think 90% of the value of this log will be in the single day after the error when recovering.

sahil839 commented:
We do have `RealmAuditLog` entries for setting changes. Can those be used to restore permissions?

timabbott commented:
We can, but it'd be nice to have a single object that records the information -- otherwise, you end up having to look through and try to find all the `RealmAuditLog` changes to settings from about the same time. 

(At least, if it's not too messy to record).

sahil839 commented on zerver/lib/events.py:
```diff
@@ -1208,22 +1207,6 @@ def _draft_update_action(i: int) -> None:
                             if user_id != person_user_id
                         ]
-
-                    for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS:
```
Previously, we send setting update events only for guests who cannot access the user. For all other users, clients were responsible to update the settings on receiving `update` event.

But the setting value did not change previously, now setting value is actually changed as we are removing the deactivated user from group. So, have updated the code to send events for setting updates for all users.

sahil839 commented on zerver/tests/test_events.py:
```diff
@@ -3414,59 +3396,19 @@ def test_do_deactivate_user(self) -> None:
         check_user_group_remove_members("events[0]", events[0])
         check_user_group_remove_members("events[1]", events[1])
         check_user_group_remove_members("events[2]", events[2])
-        check_user_group_update("events[3]", events[3], {"can_add_members_group"})
```
Removed this since we have a separate test testing this.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -525,8 +701,10 @@ def do_deactivate_user(
             event_time=event_time,
             extra_data={
                 RealmAuditLog.ROLE_COUNT: realm_user_count_by_role(user_profile.realm),
+                "removed_permissions": removed_permissions_audit_log_data,
```
Not sure if this field name is fine. The value here is a list of objects with one object per setting. For stream and groups, each setting for a stream or group is a separate object.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -498,7 +654,15 @@ def do_deactivate_user(
         bot_profiles = get_active_bots_owned_by_user(user_profile)
         for profile in bot_profiles:
-            do_deactivate_user(profile, _cascade=False, acting_user=acting_user)
+            bot_group_setting_updates = check_group_permission_updates_for_deactivating_user(
```
This will result in queries in a loop. It also happens in `delete_user` management command which allows deleting multiple users. Still working on updating this to use a bulk query.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -280,108 +293,6 @@ def change_user_is_active(user_profile: UserProfile, value: bool) -> None:
-                setting_group,
-                group_members,
-                group_subgroups,
-                notify_user_ids,
```
Added a new function for sending these events since we now have data for all the settings in `GroupPermissionUpdates`. Also, all realm settings are sent in a single event unlike what we did here with one event for each setting.

sahil839 commented:
Regarding the optimizing part, it is easier to have bulk version of `check_group_permission_updates_for_deactivating_users` to fetch `GroupPermissionUpdates` of multiple users in O(1) number of queries.

But it will be somewhat ugly to have a bulk version of `update_group_settings_for_deactivated_user` so that setting can be updated considering all the users being deactivated together. It becomes somewhat problematic because we deactivate each user separately, like when we deactivate bots owned by a user being deactivated. One option can be to have a parameter passed to `do_deactivate_user` which will decide whether to udpate the settings in `do_deactivate_user` or in the function calling `do_deactivate_user`.

shubham-padia reviewed (commented):
Haven't completed the review, but posting some initial comments if it helps unblock some work.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -1222,3 +1238,151 @@ def max_message_id_for_user(user_profile: UserProfile | None) -> int:
+
+    user_anonymous_group_ids = UserGroupMembership.objects.filter(
+        user_profile=user_profile, user_group__named_user_group=None
+    ).values_list("user_group_id", flat=True)
```
Even if anonymous groups could be subgroups, they would have no affect as we want to remove a user from a group. Removing a user from its subgroup would just mean the user also does not belong to the supergroup. So, am not sure if we need a comment here.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -511,14 +667,34 @@ def do_deactivate_user(
+        if group_setting_updates is None:
+            # For most cases, callers should pass this. This is added mainly
+            # for tests, so that we do not need to pass this repeatedly when
+            # testing deactivated users.
```
Have done this to maintain the pattern that we follow in most places where we make sure all the checks are made before calling a function in `actions` module.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -496,8 +393,267 @@ def send_events_for_user_deactivation(user_profile: UserProfile) -> None:
+    event_time = timezone_now()
+    realm_audit_logs_objects = []
+    realm_event_data = {}
+    for realm_setting_update in group_permission_updates.realm_settings_to_update:
```
I tried but have not been able to refactor it in a way that makes this more readable as the lists updated in this loop are used further and the new setting value computed is needed for event and audit log data. Any suggestions from your side would be helpful.

shubham-padia replied:
I spent a bit of time on this and agreed that it's hard to find a way that improves readability. The following attempt helps solve the duplication problem between all loops and maybe improves the readability a bit. 

I'll leave it upto you to try this patch and see if the existing code or the patched code makes more sense to have. There might be changes required to the patch in the form of better variable names and some minor changes if the approach sounds good to you.

```
diff --git a/zerver/actions/users.py b/zerver/actions/users.py
index 876521ffba..b7359e8fe7 100644
--- a/zerver/actions/users.py
+++ b/zerver/actions/users.py
@@ -1,7 +1,7 @@
 import secrets
 from collections import defaultdict
 from email.headerregistry import Address
-from typing import Any
+from typing import Any, NamedTuple
 
 from django.conf import settings
 from django.contrib.auth.tokens import PasswordResetTokenGenerator, default_token_generator
@@ -52,6 +52,7 @@ from zerver.lib.user_groups import (
 )
 from zerver.lib.users import (
     GroupPermissionUpdates,
+    SettingToUpdate,
     check_group_permission_updates_for_deactivating_user,
     get_active_bots_owned_by_user,
     get_user_ids_who_can_access_user,
@@ -418,64 +419,106 @@ def get_new_setting_value_after_deactivation(
     )
 
 
-def update_group_settings_for_deactivated_user(
-    group_permission_updates: GroupPermissionUpdates,
+class PermissionSettingChangeResult(NamedTuple):
+    value_changed: bool
+    new_user_group_members_dict: int | UserGroupMembersDict
+
+
+def _apply_permission_setting_change(
+    setting_object: Realm | Stream | NamedUserGroup,
+    setting_update: SettingToUpdate,
     user_profile: UserProfile,
     acting_user: UserProfile | None,
-) -> None:
-    anonymous_groups_to_delete = []
-    anonymous_groups_to_update = []
-
+    event_type: AuditLogEventType,
+    update_settings_list: set[str],
+    anonymous_groups_to_delete: list[int],
+    anonymous_groups_to_update: list[int],
+    realm_audit_logs_objects: list[RealmAuditLog],
+) -> PermissionSettingChangeResult:
     realm = user_profile.realm
+    setting_name = setting_update.setting_name
+    old_setting_value = setting_update.group_setting_value
     nobody_group = NamedUserGroup.objects.get(
         name=SystemGroups.NOBODY, realm=realm, is_system_group=True
     )
+    anonymous_group_id = getattr(setting_object, setting_name + "_id")
+    new_setting_value = get_new_setting_value_after_deactivation(
+        old_setting_value, user_profile, nobody_group
+    )
 
-    updated_realm_settings = []
-
-    all_active_user_ids = active_user_ids(realm.id)
+    value_changed = isinstance(new_setting_value, int)
+    if value_changed:
+        anonymous_groups_to_delete.append(anonymous_group_id)
+        setattr(setting_object, setting_name + "_id", new_setting_value)
+        update_settings_list.add(setting_name)
+    else:
+        anonymous_groups_to_update.append(anonymous_group_id)
 
-    event_time = timezone_now()
-    realm_audit_logs_objects = []
-    realm_event_data = {}
-    for realm_setting_update in group_permission_updates.realm_settings_to_update:
-        setting_name = realm_setting_update.setting_name
-        old_setting_value = realm_setting_update.group_setting_value
-        anonymous_group_id = getattr(realm, setting_name + "_id")
+    if isinstance(setting_object, Stream):
+        audit_log_kwargs: dict[str, Any] = dict(modified_stream=setting_object)
+    elif isinstance(setting_object, NamedUserGroup):
+        audit_log_kwargs = dict(modified_user_group=setting_object)
+    else:
+        audit_log_kwargs = {}
 
-        new_setting_value = get_new_setting_value_after_deactivation(
-            old_setting_value, user_profile, nobody_group
+    realm_audit_logs_objects.append(
+        RealmAuditLog(
+            realm=realm,
+            acting_user=acting_user,
+            event_type=event_type,
+            event_time=timezone_now(),
+            extra_data={
+                RealmAuditLog.OLD_VALUE: get_group_setting_value_for_audit_log_data(
+                    old_setting_value
+                ),
+                RealmAuditLog.NEW_VALUE: get_group_setting_value_for_audit_log_data(
+                    new_setting_value
+                ),
+                "property": setting_name,
+            },
+            **audit_log_kwargs,
         )
+    )
 
-        if isinstance(new_setting_value, int):
-            anonymous_groups_to_delete.append(anonymous_group_id)
-            setattr(realm, setting_name + "_id", new_setting_value)
-            updated_realm_settings.append(setting_name)
-        else:
-            anonymous_groups_to_update.append(anonymous_group_id)
+    new_user_group_members_dict = convert_to_user_group_members_dict(new_setting_value)
 
-        realm_audit_logs_objects.append(
-            RealmAuditLog(
-                realm=realm,
-                event_type=AuditLogEventType.REALM_PROPERTY_CHANGED,
-                event_time=event_time,
-                acting_user=acting_user,
-                extra_data={
-                    RealmAuditLog.OLD_VALUE: get_group_setting_value_for_audit_log_data(
-                        old_setting_value
-                    ),
-                    RealmAuditLog.NEW_VALUE: get_group_setting_value_for_audit_log_data(
-                        new_setting_value
-                    ),
-                    "property": setting_name,
-                },
-            )
+    return PermissionSettingChangeResult(value_changed, new_user_group_members_dict)
+
+
+def update_group_settings_for_deactivated_user(
+    group_permission_updates: GroupPermissionUpdates,
+    user_profile: UserProfile,
+    acting_user: UserProfile | None,
+) -> None:
+    anonymous_groups_to_delete: list[int] = []
+    anonymous_groups_to_update: list[int] = []
+    realm_audit_logs_objects: list[RealmAuditLog] = []
+
+    realm = user_profile.realm
+    all_active_user_ids = active_user_ids(realm.id)
+
+    updated_realm_settings: set[str] = set()
+    realm_event_data: dict[str, int | UserGroupMembersDict] = {}
+
+    for realm_setting_update in group_permission_updates.realm_settings_to_update:
+        change_result = _apply_permission_setting_change(
+            setting_object=realm,
+            setting_update=realm_setting_update,
+            user_profile=user_profile,
+            acting_user=acting_user,
+            event_type=AuditLogEventType.REALM_PROPERTY_CHANGED,
+            update_settings_list=updated_realm_settings,
+            anonymous_groups_to_delete=anonymous_groups_to_delete,
+            anonymous_groups_to_update=anonymous_groups_to_update,
+            realm_audit_logs_objects=realm_audit_logs_objects,
         )
 
-        realm_event_data[setting_name] = convert_to_user_group_members_dict(new_setting_value)
+        realm_event_data[realm_setting_update.setting_name] = (
+            change_result.new_user_group_members_dict
+        )
 
-    if len(updated_realm_settings) != 0:
-        realm.save(update_fields=updated_realm_settings)
+    if updated_realm_settings:
+        realm.save(update_fields=list(updated_realm_settings))
 
     if realm_event_data:
         event = dict(
@@ -486,8 +529,8 @@ def update_group_settings_for_deactivated_user(
         )
         send_event_on_commit(realm, event, all_active_user_ids)
 
-    updated_stream_settings = set()
-    stream_objs_to_update = []
+    updated_stream_settings: set[str] = set()
+    stream_objs_to_update: list[Stream] = []
 
     updated_streams = [
         stream_setting_update.stream
@@ -500,47 +543,27 @@ def update_group_settings_for_deactivated_user(
         setting_value_changed = False
 
         for setting_update in stream_setting_update.settings:
-            setting_name = setting_update.setting_name
-            old_setting_value = setting_update.group_setting_value
-            anonymous_group_id = getattr(stream, setting_name + "_id")
-
-            new_setting_value = get_new_setting_value_after_deactivation(
-                old_setting_value, user_profile, nobody_group
+            change_result = _apply_permission_setting_change(
+                setting_object=stream,
+                setting_update=setting_update,
+                user_profile=user_profile,
+                acting_user=acting_user,
+                event_type=AuditLogEventType.CHANNEL_GROUP_BASED_SETTING_CHANGED,
+                update_settings_list=updated_stream_settings,
+                anonymous_groups_to_delete=anonymous_groups_to_delete,
+                anonymous_groups_to_update=anonymous_groups_to_update,
+                realm_audit_logs_objects=realm_audit_logs_objects,
             )
 
-            if isinstance(new_setting_value, int):
-                anonymous_groups_to_delete.append(anonymous_group_id)
-                setattr(stream, setting_name + "_id", new_setting_value)
+            if change_result.value_changed:
                 setting_value_changed = True
-                updated_stream_settings.add(setting_name)
-            else:
-                anonymous_groups_to_update.append(anonymous_group_id)
-
-            realm_audit_logs_objects.append(
-                RealmAuditLog(
-                    realm=realm,
-                    acting_user=acting_user,
-                    modified_stream=stream,
-                    event_type=AuditLogEventType.CHANNEL_GROUP_BASED_SETTING_CHANGED,
-                    event_time=event_time,
-                    extra_data={
-                        RealmAuditLog.OLD_VALUE: get_group_setting_value_for_audit_log_data(
-                            old_setting_value
-                        ),
-                        RealmAuditLog.NEW_VALUE: get_group_setting_value_for_audit_log_data(
-                            new_setting_value
-                        ),
-                        "property": setting_name,
-                    },
-                )
-            )
 
             notify_user_ids = can_access_stream_ids_dict[stream.id]
             stream_event = dict(
                 op="update",
                 type="stream",
-                property=setting_name,
-                value=convert_to_user_group_members_dict(new_setting_value),
+                property=setting_update.setting_name,
+                value=change_result.new_user_group_members_dict,
                 stream_id=stream.id,
                 name=stream.name,
             )
@@ -549,60 +572,40 @@ def update_group_settings_for_deactivated_user(
         if setting_value_changed:
             stream_objs_to_update.append(stream)
 
-    if len(stream_objs_to_update) != 0:
+    if stream_objs_to_update:
         Stream.objects.bulk_update(stream_objs_to_update, fields=list(updated_stream_settings))
 
-    updated_group_settings = set()
-    group_objs_to_update = []
+    updated_group_settings: set[str] = set()
+    group_objs_to_update: list[NamedUserGroup] = []
     for group_setting_update in group_permission_updates.group_settings_to_update:
         group = group_setting_update.group
         setting_value_changed = False
         event_data: dict[str, str | int | UserGroupMembersDict] = {}
 
         for setting_update in group_setting_update.settings:
-            setting_name = setting_update.setting_name
-            old_setting_value = setting_update.group_setting_value
-            anonymous_group_id = getattr(group, setting_name + "_id")
-
-            new_setting_value = get_new_setting_value_after_deactivation(
-                old_setting_value, user_profile, nobody_group
+            change_result = _apply_permission_setting_change(
+                setting_object=group,
+                setting_update=setting_update,
+                user_profile=user_profile,
+                acting_user=acting_user,
+                event_type=AuditLogEventType.USER_GROUP_GROUP_BASED_SETTING_CHANGED,
+                update_settings_list=updated_group_settings,
+                anonymous_groups_to_delete=anonymous_groups_to_delete,
+                anonymous_groups_to_update=anonymous_groups_to_update,
+                realm_audit_logs_objects=realm_audit_logs_objects,
             )
 
-            if isinstance(new_setting_value, int):
-                anonymous_groups_to_delete.append(anonymous_group_id)
-                setattr(group, setting_name + "_id", new_setting_value)
-                setting_value_changed = True
-                updated_group_settings.add(setting_name)
-            else:
-                anonymous_groups_to_update.append(anonymous_group_id)
-
-            realm_audit_logs_objects.append(
-                RealmAuditLog(
-                    realm=realm,
-                    acting_user=acting_user,
-                    event_type=AuditLogEventType.USER_GROUP_GROUP_BASED_SETTING_CHANGED,
-                    event_time=timezone_now(),
-                    modified_user_group=group,
-                    extra_data={
-                        RealmAuditLog.OLD_VALUE: get_group_setting_value_for_audit_log_data(
-                            old_setting_value
-                        ),
-                        RealmAuditLog.NEW_VALUE: get_group_setting_value_for_audit_log_data(
-                            new_setting_value
-                        ),
-                        "property": setting_name,
-                    },
-                )
-            )
+            event_data[setting_update.setting_name] = change_result.new_user_group_members_dict
 
-            event_data[setting_name] = convert_to_user_group_members_dict(new_setting_value)
+            if change_result.value_changed:
+                setting_value_changed = True
 
         do_send_user_group_update_event(group, event_data)
 
         if setting_value_changed:
             group_objs_to_update.append(group)
 
-    if len(group_objs_to_update) != 0:
+    if group_objs_to_update:
         NamedUserGroup.objects.bulk_update(
             group_objs_to_update, fields=list(updated_group_settings)
         )
```

sahil839 replied:
I am not sure about this since we need to pass a whole lot of parameters to `_apply_permission_setting_change`. Have added this in a separate commit for now to get feedback from Tim as well and then we can squash if the refactor is fine.

shubham-padia replied:
sounds good.

shubham-padia commented on zerver/tests/test_events.py:
```diff
@@ -4357,6 +4245,122 @@ def test_notification_setting_event_not_sent(self) -> None:
+            hamletcharacters_group,
+            "can_manage_group",
+            setting_group,
+            acting_user=None,
```
I think some comments would be helpful in all the test setup happening above. Hard to tell quickly why what setup is happening.

shubham-padia commented on zerver/views/users.py:
```diff
@@ -144,7 +145,11 @@ def deactivate_user_own_backend(request: HttpRequest, user_profile: UserProfile)
         raise CannotDeactivateLastUserError(is_last_owner=True)
 
-    do_deactivate_user(user_profile, acting_user=user_profile)
+    group_setting_updates = check_group_permission_updates_for_deactivating_user(user_profile)
```
We have error handling for this function everywhere. Is there some error handling we need to do in views as well?

sahil839 replied:
For other places, we cannot raise a `JsonableError` directly, like for managements command we raise `CommandError` and for support views we return a success response with context containing error message.

shubham-padia reviewed (commented):
Mostly looks good now. Added some minor comments.

sahil839 committed:
(squash) users: Deduplicate code in update_group_settings_for_deactivated_user.

shubham-padia reviewed (commented):
Looks good to me except one doubt. How have we tested the migration here? A lot is going on in the migration and I'm not confident that I've missed something or not.

timabbott commented:
I can try to merge this tomorrow if you can get it rebased.

sahil839 commented:
Rebased the PR.

mateuszmandera commented on zerver/lib/users.py:
```diff
@@ -1203,3 +1220,147 @@ def max_message_id_for_user(user_profile: UserProfile | None) -> int:
+
+    return settings_to_update, objection_settings
+
+
+def check_group_permission_updates_for_deactivating_user(
+    user_profile: UserProfile,
+) -> GroupPermissionUpdates:
```
I think this function really needs a docstring explaining what the checking logic is and how/when callers should use it (will be helpful for me to understand and review the PR as well).

I also have some concerns about allowing this to block actions such as user deletion via the support panel and CLI. I'll post in the #backend thread tomorrow after giving this a bit more thought.

sahil839 replied:
Added a docstring. Let me know if it needs some more details.

sahil839 committed:
settings: Add replacement_group_name to GroupPermissionSetting.

sahil839 committed:
settings: Add migration to remove deactivated users from anonymous groups.

This commit adds migration to remove already deactivated users from
existing anonymous groups being used for different permission settings
to maintain the behavior we added in the previous commit of removing
the user from anonymous groups while deactivate them.

We also backfill the audit log entries for setting values being
changed.

sahil839 commented:
Updated the PR as discussed on CZO.
- We force deactivate user when using management commands, support view or LDAP/SCIM sync. For settings that cannot be set to nobody group, they are set to `replacement_group_name` (added in the first commit of this PR) which is owners group.
- Deactivated user is removed from settings of deactivated channels and groups as well.
